fix: Delay test hook deletion until after logs are retrieved with `helm test release --logs`

benoittgt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

This is a rebase of #10603

Fixing a [very old issue](https://github.com/helm/helm/pull/10603) with frequent activity from users.

When running `helm test <release-name> --logs`, test pods with `hook-succeeded` delete policy were being deleted before logs could be retrieved, resulting in "pods not found" errors. This PR delays the execution of hook deletion until after logs have been successfully retrieved.

**Special notes for your reviewer**:

Tested with a chart 
1. `helm create test-chart`
2. Modify the `tests/test-connection.yaml` generated file by adding one annotation `"helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded`
3. Run test
  ```
  helm test new-chart --logs        
NAME: new-chart
LAST DEPLOYED: Wed Nov 26 17:18:45 2025
NAMESPACE: default
STATUS: deployed
REVISION: 2
DESCRIPTION: Upgrade complete
TEST SUITE:     new-chart-test-connection
Last Started:   Wed Nov 26 17:18:48 2025
Last Completed: Wed Nov 26 17:18:51 2025
Phase:          Succeeded

Error: unable to get pod logs for new-chart-test-connection: pods "new-chart-test-connection" not found
```
4. With this branch
```
bin/helm test new-chart --logs
NAME: new-chart
LAST DEPLOYED: Wed Nov 26 17:18:45 2025
NAMESPACE: default
STATUS: deployed
REVISION: 2
DESCRIPTION: Upgrade complete
TEST SUITE:     new-chart-test-connection
Last Started:   Wed Nov 26 17:20:04 2025
Last Completed: Wed Nov 26 17:20:08 2025
Phase:          Succeeded

POD LOGS: new-chart-test-connection
Connecting to new-chart:80 (192.168.194.168:80)
saving to 'index.html'
index.html           100% |********************************|   612  0:00:00 ETA
'index.html' saved
``` 

Fixes #9098
Closes: #10603

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

mfbieber committed:
added shutdown hook that is executed after the logs have been retrieved

Signed-off-by: Michelle Fernandez Bieber <mfbieber@gmail.com>

mfbieber committed:
updated comment and made defer of shutdown function return errors as before and not the possible shutdown error

Signed-off-by: Michelle Fernandez Bieber <mfbieber@gmail.com>

mfbieber committed:
cleaned up empty line

Signed-off-by: Michelle Fernandez Bieber <mfbieber@gmail.com>

mfbieber committed:
added check for nil shutdown

Signed-off-by: Michelle Fernandez Bieber <mfbieber@gmail.com>

mfbieber committed:
Update pkg/action/hooks.go

Co-authored-by: Marco Lecheler <marco@task.media>
Signed-off-by: Michelle Fernandez Bieber <37021266+mfbieber@users.noreply.github.com>

benoittgt committed:
Fix linting issue

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

banjoh reviewed (commented):
Changes look good overall. I left a few comments

banjoh commented on pkg/action/hooks.go:
```diff
@@ -33,6 +33,27 @@ import (
+	return shutdown()
+}
+
+type ExecuteShutdownHooks = func() error
```
nit pick: Rename to executeShutdownFunc to depict its purpose as a function

benoittgt replied:
Linting is not happy

```
pkg/action/release_testing.go:60:57: unexported-return: exported method Run returns unexported type action.executeShutdownFunc, which can be annoying to use (revive)
func (r *ReleaseTesting) Run(name string) (ri.Releaser, executeShutdownFunc, error) {
```

benoittgt committed:
Code review

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Deal with golint warning with private executeShutdownFunc

```
  Error: pkg/action/release_testing.go:60:57: unexported-return: exported method Run returns unexported type action.executeShutdownFunc, which can be annoying to use (revive)
  func (r *ReleaseTesting) Run(name string) (ri.Releaser, executeShutdownFunc, error) {

```

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

banjoh reviewed (approved):
LGMT

TerryHowe reviewed (approved):
/lgtm
